src/history/test/HistoryTestsUtils.cpp

```diff
@@ -489,74 +484,58 @@ CatchupSimulation::generateRandomLedger()
 }
 
 void
-CatchupSimulation::generateAndPublishHistory(size_t nPublishes)
+CatchupSimulation::ensureLedgerAvailable(uint32_t targetLedger)
 {
     auto& lm = mApp.getLedgerManager();
     auto& hm = mApp.getHistoryManager();
-
-    size_t publishSuccesses = hm.getPublishSuccessCount();
-    SequenceNumber ledgerSeq = lm.getLastClosedLedgerNum() + 1;
-
-    while (hm.getPublishSuccessCount() < (publishSuccesses + nPublishes))
+    while (lm.getLastClosedLedgerNum() < targetLedger)
     {
-        uint64_t queueCount = hm.getPublishQueueCount();
-        while (hm.getPublishQueueCount() == queueCount)
-        {
-            generateRandomLedger();
-            ++ledgerSeq;
-        }
-
-        mBucketListAtLastPublish = getApp().getBucketManager().getBucketList();
-
-        // One more ledger is needed to close as stellar-core only publishes
-        // to just-before-LCL
-        generateRandomLedger();
-        ++ledgerSeq;
-        // One more for trigger ledger
         generateRandomLedger();
-        REQUIRE(mApp.getLedgerManager().getLastClosedLedgerNum() == ledgerSeq);
-        ++ledgerSeq;
 
-        // Advance until we've published (or failed to!)
-        while (hm.getPublishSuccessCount() < hm.getPublishQueueCount())
+        auto seq = mApp.getLedgerManager().getLastClosedLedgerNum() + 1;
+        if (seq == hm.nextCheckpointLedger(seq))
         {
-            REQUIRE(hm.getPublishFailureCount() == 0);
-            mApp.getClock().crank(true);
+            mBucketListAtLastPublish =
+                getApp().getBucketManager().getBucketList();
         }
     }
+}
+
+void
+CatchupSimulation::ensurePublishesComplete()
+{
+    auto& hm = mApp.getHistoryManager();
+    while (!mApp.getWorkManager().allChildrenDone() ||
+           (hm.getPublishSuccessCount() < hm.getPublishQueueCount()))
+    {
+        REQUIRE(hm.getPublishFailureCount() == 0);
+        mApp.getClock().crank(true);
+    }
 
     REQUIRE(hm.getPublishFailureCount() == 0);
-    REQUIRE(hm.getPublishSuccessCount() == publishSuccesses + nPublishes);
-    REQUIRE(mApp.getLedgerManager().getLastClosedLedgerNum() ==
-            ((publishSuccesses + nPublishes) * hm.getCheckpointFrequency()) +
-                1);
 }
 
-Application::pointer
-CatchupSimulation::catchupNewApplication(uint32_t initLedger, uint32_t count,
-                                         bool manual, Config::TestDbMode dbMode,
-                                         std::string const& appName)
+void
+CatchupSimulation::ensureOfflineCatchupPossible(uint32_t targetLedger)
```

Comment:
Yes, but that would make tests harder to understand. Instead of asking 'why do I need to make ledgers XYZ available' you can now look at test and see, 'oh, so from now it should be possible to make online/offline catchup to the ledger I've just passed to that method as a parameter'.